src/rpc/rawtransaction.cpp

```diff
@@ -822,20 +822,27 @@ UniValue signrawtransaction(const JSONRPCRequest& request)
                 view.AddCoin(out, std::move(newcoin), true);
             }
 
-            // if redeemScript given and not using the local wallet (private keys
-            // given), add redeemScript to the tempKeystore so it can be signed:
-            if (fGivenKeys && (scriptPubKey.IsPayToScriptHash() || scriptPubKey.IsPayToWitnessScriptHash())) {
-                UniValue rs = find_value(prevOut, "redeemScript");
-                std::vector<UniValue> scripts;
-                if (rs.getType() == UniValue::VSTR) {
-                    scripts.push_back(rs);
-                } else if (rs.getType() == UniValue::VARR) {
+            // if redeemScript given add redeemScript to the tempKeystore so it can be signed:
+            UniValue rs = find_value(prevOut, "redeemScript");
+            if (rs.getType() != UniValue::VNULL) {
+                if (!(scriptPubKey.IsPayToScriptHash() || scriptPubKey.IsPayToWitnessScriptHash())) {
+                    // if this isnt P2SH or P2WSH, scripts should not be provided
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Scripts should not be provided if P2SH or P2WSH are not being used");
+                } else if (!fGivenKeys) {
+                    // if private keys aren't given, assume we are using the local wallet, so we should not accept scripts
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Scripts cannot be provided when using local wallet (no keys given)");
+                }
+                std::vector<std::string> scripts;
+                if (rs.getType() == UniValue::VARR) {
                     // If redeemScript is an array, iterate over all scripts provided
-                    scripts = rs.getValues();
+                    for (auto r : rs.getValues()) {
+                        scripts.push_back(r.get_str());
+                    }
+                } else {
+                    scripts.push_back(rs.get_str());
                 }
-                for (const UniValue& script : scripts) {
-                    if (script.getType() != UniValue::VSTR) continue;
-                    std::vector<unsigned char> script_data(ParseHexV(script.get_str(), "redeemScript"));
+                for (const std::string script : scripts) {
```

Comment:
Could use const reference to avoid string copy here (though there are more copies below anyway).